all: Prefer ss::coroutine::maybe_yield

StephanDollberg opened the pull request:
When used in a coroutine context `ss::coroutine::maybe_yield` saves a
couple hundred instructions over `ss::maybe_yield` as the whole `future`
machinery isn't getting involved.

Can be seen in the following godbolt: https://godbolt.org/z/Wdb441PYE

X-Coauthored-By: Travis Downs <travis.downs@redpanda.com>

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.3.x
- [ ] v23.2.x
- [ ] v23.1.x

## Release Notes

* none

emaxerrno commented:
@StephanDollberg i guess we could do the alias on `ssx` namespace and always do a blind-replace?

rockwotj reviewed (dismissed):
LGTM - it's a good candidate for a lint check, when we have better tooling in place for simple lint checks.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/44104#018d3297-e9a7-44db-b268-090f5a2704df
<!--d9271559-1863-42fe-8934-3192caf6b8df-->

StephanDollberg commented:
/ci-repeat
debug
skip-units
dt-repeat=15
tests/rptest/tests/cpu_stress_injection_test.py

StephanDollberg commented:
/ci-repeat
release
skip-units
dt-repeat=15
tests/rptest/tests/cpu_stress_injection_test.py

StephanDollberg commented:
> @StephanDollberg i guess we could do the alias on ssx namespace and always do a blind-replace?

Maybe not sure I fully understand but what would putting it into ssx get us in that respect? People would still have to remember to use `ssx::maybe_yield` over `ss::maybe_yield` so I don't think there is all that much benefit like for the other ssx stuff where we actually add some functionality.

 > LGTM - it's a good candidate for a lint check, when we have better tooling in place for simple lint checks.

Yeah linter would be good but at the same time it's not super important to write some extra C++.

So actually there is one more detail that we originally missed. `ss::maybe_yield` is also less tight than `ss::coroutine::maybe_yield`. The former takes two loops through the task queue until the original task gets scheduled again. The later just one. This made the stress fiber test fail. We did realize then though that that is already flaky (https://github.com/redpanda-data/redpanda/issues/13701) which is caused by the stress fibers actually running in the admin scheduling group and hence actually starving the admin api server.

To work around that I have moved the stress fibers into the main group and made the scheduling group also configurable.

cc @andrwng

andrwng commented on src/v/finjector/stress_fiber.cc:
```diff
@@ -17,11 +17,15 @@
 class stress_payload {
 public:
-    // Starts the stress payload with the given configuration.
+    // Starts the stress payload with the given configuration in the given
```
nit: would be great if this could be pulled out into a separate commit. You never know when it's useful to cherry-pick something, and it'd be nice to not have to pull in the wider maybe_yield change if we do (though tbf the scheduling group change is small)

StephanDollberg replied:
Yes good point, done

andrwng commented on src/v/finjector/stress_fiber.h:
```diff
@@ -27,7 +28,11 @@ struct stress_config {
     size_t num_fibers;
+
+    // the seastar scheduling group to run the stress fibers in
+    ss::scheduling_group sg;
```
Nice!

andrwng reviewed (approved):
Nice find! Just so I understand, will `couroutine::maybe_yield` inherently be the same or faster because it fundamentally has less work to do by avoiding the future machinery? Are the two loops through the task queue specific to coroutines vs futures, or just a difference in implementation?

StephanDollberg committed:
test: Support setting scheduling group in stress fiber

The `cpu_stress_injection_test` test was occasionally flaky. This was
because the stress fiber runs in the admin scheduling group (as that is
where it's started from). This however starves the admin server itself
and hence it can time out sending the actual API response.

To work around that issue we move the stress fibers to run in the main
scheduling group when invoked from the admin api.

To do that we extend the stress fiber api such that the scheduling group
can be specified.

Fixes https://github.com/redpanda-data/redpanda/issues/13701

StephanDollberg committed:
all: Prefer ss::coroutine::maybe_yield

When used in a coroutine context `ss::coroutine::maybe_yield` saves a
couple hundred instructions over `ss::maybe_yield` as the whole `future`
machinery isn't getting involved.

Can be seen in the following godbolt: https://godbolt.org/z/Wdb441PYE

There is another difference where `ss::coroutine::maybe_yield` will only
cause one loop through the task queue while `ss::maybe_yield` requires
two. This is because `ss::maybe_yield` is implemented by waiting for an
empty task to resolve which requires one run through the task queue for
the empty task to run (which marks the future the original task waits on
ready and then enqueues the original task back into the task queue) and
a second to then yield back to the original task. The coroutine version
yields back directly. This makes the coroutine version a little bit
tighter.

Fixes https://github.com/redpanda-data/redpanda/issues/13701

Co-authored-by: Travis Downs <travis.downs@redpanda.com>

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/44285#018d4046-0fad-49c6-a507-38cd92c6518c
<!--9fd305cb-bf0e-474e-a736-6cd8d6535f4b-->

andrwng commented:
> > Nice find! Just so I understand, will couroutine::maybe_yield inherently be the same or faster because it fundamentally has less work to do by avoiding the future machinery?
> 
> Yes exactly
> 
> > Are the two loops through the task queue specific to coroutines vs futures, or just a difference in implementation?
> 
> No I don't think that's fundamental to coroutines vs futures. It's more an implementation detail of yield I think where in the coroutine case you get a future to "yourself" immediately on suspend. But I don't think this translates to all cases in regards to coroutines vs futures.

Got it, thank you for the explanation!

piyushredpanda commented:
Feel free to merge if you are good with it, @StephanDollberg
